azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/functions/DeploymentToNodeMetadata.java

```diff
@@ -0,0 +1,162 @@
+      if (text != null) {
+         for (Deployment.ProvisioningState status : Deployment.ProvisioningState.values()) {
+            if (text.equalsIgnoreCase(status.name())) {
+               return status;
+            }
+         }
+      }
+      return Deployment.ProvisioningState.UNRECOGNIZED;
+   }
+
+   private final AzureComputeApi api;
+
+   private final Supplier<Set<? extends Location>> locations;
+
+   private final GroupNamingConvention nodeNamingConvention;
+
+   private final VMImageToImage vmImageToImage;
+
+   private final VMHardwareToHardware vmHardwareToHardware;
+
+   private final Map<String, Credentials> credentialStore;
+
+   @Inject
+   DeploymentToNodeMetadata(
+           AzureComputeApi api,
+           @Memoized Supplier<Set<? extends Location>> locations,
+           GroupNamingConvention.Factory namingConvention, VMImageToImage vmImageToImage,
+           VMHardwareToHardware vmHardwareToHardware, Map<String, Credentials> credentialStore) {
+
+      this.nodeNamingConvention = namingConvention.createWithoutPrefix();
+      this.locations = locations;
+      this.vmImageToImage = vmImageToImage;
+      this.vmHardwareToHardware = vmHardwareToHardware;
+      this.credentialStore = credentialStore;
+      this.api = api;
+   }
+
+   @Override
+   public NodeMetadata apply(final VMDeployment from) {
+      final NodeMetadataBuilder builder = new NodeMetadataBuilder();
+      final Deployment deployment = from.deployment;
+      builder.id(deployment.name());
+      builder.providerId(deployment.name());
+      builder.name(deployment.name());
+      String group = deployment.name();
+      int index = group.lastIndexOf("-");
+      builder.group(group.substring(0, index));
+
+      NodeMetadata.Status status = STATUS_TO_NODESTATUS.get(provisioningStateFromString(deployment.properties().provisioningState()));
+      if (status == NodeMetadata.Status.RUNNING && from.vm != null && from.vm.statuses() != null) {
+         List<VirtualMachineInstance.VirtualMachineStatus> statuses = from.vm.statuses();
+         for (int c = 0; c < statuses.size(); c++) {
+            if (statuses.get(c).code().substring(0, 10).equals("PowerState")) {
+               if (statuses.get(c).displayStatus().equals("VM running")) {
+                  status = NodeMetadata.Status.RUNNING;
+               } else if (statuses.get(c).displayStatus().equals("VM stopped")) {
+                  status = NodeMetadata.Status.SUSPENDED;
+               }
+               break;
+            }
+         }
+      }
+
+      builder.status(status);
+
+      Credentials credentials = credentialStore.get("node#" + from.deployment.name());
+      if (credentials != null && credentials.identity.equals(JCLOUDS_DEFAULT_USERNAME)) {
+         credentials = new Credentials(AZURE_DEFAULT_USERNAME, credentials.credential);
+      } else if (credentials == null) {
+         credentials = new Credentials(AZURE_DEFAULT_USERNAME, DEFAULT_LOGIN_PASSWORD);
+      }
```

Comment:
So, the question here is "what happens if a user sets an invalid username when configuring the TemplateOptions?"
* If the deployment fails, then it is OK.
* If the deployment succeeds, then which username does the VM get?

In any case, jclouds has a mechanism to provide "default" credentials for images. Those default image credentials could be populated to the node if the user does not explicitly set ones.
Instead of setting the credentials constants in the DeploymentTemplateBuilder, you could configurate jclouds to populate them as default image credentials, then they could be accessed from a template with just `tempalte.getDefaultCredentials()` and be read by the DeploymentTemplateBuilder to provide the fallback credentials.

Configuring the default credentials for images just requires to set the `ComputeServiceProperties.IMAGE_LOGIN_USER` property to a sane default value for the username, or to a colon-delimited form providing also a default password: `<username>:<password>`.

If you configure that property in the metadata and read it in the deployment template builder to set it if the user does not explicitly set the credentials, then you could remove this code and assume the node has proper credentials. (This will however rely on the answer to the second point above).